lib/Sema/CSDiag.cpp

```diff
@@ -1757,8 +1757,30 @@ static bool diagnoseTypeRequirementFailure(ConstraintSystem &cs,
 
   case ConstraintKind::Equal: { // same type
     if (auto *UDE = dyn_cast<UnresolvedDotExpr>(anchor)) {
-      TC.diagnose(UDE->getLoc(), diag::types_not_equal_in_call, ownerType, lhs,
-                  rhs, UDE->getName());
+      // If we actually resolved the member to use, use it.
+      auto loc = cs.getConstraintLocator(UDE, ConstraintLocator::Member);
+      if (auto *member = cs.findResolvedMemberRef(loc)) {
+        auto baseType = member->getDeclContext()->getDeclaredInterfaceType();
+        auto lhsType = constraint->getFirstType();
+        auto rhsType = constraint->getSecondType()->getAs<DependentMemberType>();
+        if (rhsType && rhsType->getBase()->isEqual(lhsType)) {
+          auto memberType = DependentMemberType::get(lhs, rhsType->getAssocType());
+          if (cs.simplifyType(memberType)->isEqual(rhs)) {
+            TC.diagnose(UDE->getLoc(), diag::types_not_equal_in_call, baseType, lhs,
+                        rhs, UDE->getName());
+          } else {
+            TC.diagnose(UDE->getLoc(), diag::types_not_equal_in_call_dependent, baseType,
+                        lhs, memberType, rhs, UDE->getName());
+          }
+        } else {
+          TC.diagnose(UDE->getLoc(), diag::types_not_equal_in_call, baseType, lhs,
+                      rhs, UDE->getName());
+        }
+        TC.diagnose(member, diag::decl_declared_here, member->getFullName());
+      } else {
+        TC.diagnose(UDE->getLoc(), diag::types_not_equal_in_call, ownerType, lhs,
+                    rhs, UDE->getName());
+      }
```

Comment:
This is… not as clean as I'd like. I'm very open to suggestions about how it might be improved.

Reply:
At least to me new messages don't seem to be much clearer about what is going on especially  for`.popFirst()`, maybe a better solution here would be to print the actual requirement which got broken like we do in other cases e.g. https://github.com/apple/swift/blob/master/test/Constraints/generics.swift#L251, WDYT?

Reply:
Hmm… it does seem like it be nice to have parity there 🤔 

Unfortunately, I'm not sure it's possible without a somewhat major change to this diagnostic. (I'd love to be wrong on this.)

That information is available on the `Requirement`. I can get to that through `member`'s `GenericContext`. It looks like the `Constraint` is generated from the `Requirement`, but I don't see a way to connect that `Requirement` back to the current `Constraint` to know which of the `Requirement`s failed.

I tried moving the `diagnoseArgumentGenericRequirements` earlier—to be before this check—but it didn't hit.

It sure seems like there ought to be a way to unify these two things. But there's quite a bit of code between them, and I don't understand enough of the big picture to know how to connect them.

This PR seems like an improvement to me, so I'd probably take this improvement for now. But I'm happy to spend some more time looking—especially if you have any suggestions.

Reply:
Constraint locator should tell you exactly what requirement is this constraint related to and you can get it by index from generic signature of the context, take a look at how `open{Unbound}Generic`/`openGenericRequirements` generates such constraints.

Reply:
How does this look?

```
error: value of type '[Int]' has no member 'popFirst'
var b = a.popFirst()
          ^
Swift.Collection:2:37: note: candidate requires that the types '[Int]' and 'ArraySlice<Int>' be equivalent (requirement specified as 'Self' == 'Self.SubSequence')
    @inlinable public mutating func popFirst() -> Self.Element?
                                    ^
```

Reply:
@mdiep I'm not sure what is going on but for some reason I can't see any new changes... But the new node looks great though!